gguf-py/gguf/metadata.py

```diff
@@ -284,20 +284,64 @@ def apply_metadata_heuristic(metadata: Metadata, model_card: Optional[dict] = No
         ########################
         if model_card is not None:
 
-            if "model_name" in model_card and metadata.name is None:
-                # Not part of huggingface model card standard but notice some model creator using it
-                # such as TheBloke in 'TheBloke/Mistral-7B-Instruct-v0.2-GGUF'
-                metadata.name = model_card.get("model_name")
+            def use_model_card_metadata(metadata: Any | None, key_name: str):
+                if key_name in model_card and metadata is None:
+                    metadata = model_card.get(key_name)
```

Comment:
Does this work? If `metadata` is `None` here, then I'm not sure changing its value would persist when exiting this function.

You'd need `metadata` to be something passed by reference (which `None` is not).

Reply:
Ah yeah, it's an object, so will need to use getattr and setattr.

The python unit test caught it and after this fix i got all passes as per below

```
python -m unittest discover ./gguf-py -v
test_apply_metadata_heuristic_from_hf_parameters (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_card (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_dir (tests.test_metadata.TestMetadataMethod) ... ok
test_get_model_id_components (tests.test_metadata.TestMetadataMethod) ... ok
test_id_to_title (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_hf_parameters (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_card (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_dir (tests.test_metadata.TestMetadataMethod) ... ok
test_get_model_id_components (tests.test_metadata.TestMetadataMethod) ... ok
test_id_to_title (tests.test_metadata.TestMetadataMethod) ... ok

----------------------------------------------------------------------
Ran 10 tests in 0.004s

OK
```

Reply:
Thanks for the catch. Hows your take on the overall concept of this PR by the way?

Reply:
> Hows your take on the overall concept of this PR by the way?

I think it's great to let the model card be used for this, as long as the semantics expected by HuggingFace are not conflicting with how metadata is used here.

(For example I think HF shows warnings when there are unknown tags in the model card, or maybe I'm mistaking it for something else, but I *think* I saw a warning related to this at least once.)

I see you've put multiple variants of the names for the same metadata keys. I think this can be good for discoverability, but (eventually, when documenting this) a single form should be recommended (while still accepting the other variants).